Emails: Improve forward form for email comparison

AllTerrainDeveloper opened the pull request:
#### Changes proposed in this Pull Request

This change requests is adding a completely new way of adding mail forwards from the Emails comparison page.
It also prepares the frontend to add multiple forwards at once, but this is blocked by backend work.
This code changes also adds new validation in emails management to forbid the creation of repeated mailboxes when multiple mailboxes creation is allowed.

In further tasks, it would be also a possibility to extend the original component `email-forwarding-add-new.jsx` to remove the `email-forwarding-add-new-compact.jsx` which is a modification to avoid regression issues on the forward management screen.

#### Testing instructions

![image](https://user-images.githubusercontent.com/5689927/142179240-6fd1f2d4-c613-459c-b051-dff2f5c5d1e2.png)

**Test 1**
1. Go to Upgrade -> Emails
2. Add a single email forward, click add. Check that the forward is added.
3. Verify that you don't see a warning at the top of the screen about an existing email forward.
4. Assert that you have been redirected to forward management screen.

**Test 2**
1. Go to Upgrade -> Emails
2. Play around with validations:

- Add a mailbox with invalid characters (like $ or %)
- Add an invalid destination mail (wrong characters, or wrong domain)

**Remarks**
In a near future there will be a new implementation, which is to add several forwards at once. The backend isn't prepared for this task, but the frontend is being prepared for this task.

Related to 1200182182542585-as-1200414925548929

AllTerrainDeveloper committed:
Emails: List of forwards, pending validation

AllTerrainDeveloper committed:
Emails: Validate & enable add button on valid forms

AllTerrainDeveloper committed:
Emails: Added validation, global validation (repeated mailboxes). Missing to verify requests.

AllTerrainDeveloper committed:
Emails: Allow only one forward at one time.

AllTerrainDeveloper committed:
Emails: Lint issue

AllTerrainDeveloper committed:
Emails: Simplify code

daledupreez commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -23,6 +25,17 @@ function validateField( { name, value } ) {
 	}
 }
 
+function validateDuplicatedForward( { name, value }, existingForwards ) {
+	switch ( name ) {
+		case 'mailbox':
+			return existingForwards?.filter( ( t ) => t.mailbox === value ).length > 1
+				? [ 'Repeated' ]
+				: [];
+		default:
+			return [];
+	}
+}
```
I am not sure that it's a bug to forward a single mailbox to multiple recipients. Do we explicitly document that this isn't allowed? We likely need to check the back end implementation to make sure we understand the current implementation.

AllTerrainDeveloper replied:
I tried to do that thing and I got errors from backend. So that's why I tried to add this validation.

AllTerrainDeveloper replied:
For reference:
![image](https://user-images.githubusercontent.com/5689927/139814236-abb73483-82ae-4c11-bcda-54fde6814e41.png)

As you say we can add multiple recipients, but not multiple (same) forwards-mailboxes.

daledupreez replied:
:+1: Then it's not allowed and we should validate early!

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,188 @@
+			);
+		} );
+
+		this.setState( { formSubmitting: false } );
```
Should we check for a Promise (or Promises?) from the calls above? We're assuming the calls return synchronously, and I don't think that's a valid assumption.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,188 @@
+						emailForwardingCount={ totalFowards }
+						emailForwardingLimit={ emailForwardingLimit }
+					/>
+				) : null }
```
Does it make more sense to show this message when `emailForwardingLimit > 0`? I am thinking that it may be useful to show the message when there is a limit, even if the user hasn't created any forwards yet?

daledupreez commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -240,17 +241,49 @@ class EmailProvidersComparison extends Component {
+	onForwardsMailboxesChange = ( changedMailboxes ) => {
+		this.setState( { forwardingMailboxes: changedMailboxes } );
+	};
+
```
Nit: we have been grouping the various provider-specific helper functions and callbacks in the file. I'd prefer if we kept using that pattern rather than mixing them.

daledupreez commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -240,17 +241,49 @@ class EmailProvidersComparison extends Component {
+		const { forwardingMailboxes } = this.state;
+
+		const usersAreValid = areAllUsersValid( forwardingMailboxes );
+		const userCanAddEmail = hasCartDomain || canCurrentUserAddEmail( domain );
```
Is this the correct validation check? What are we using in the current email form? (I believe we don't show email forwards when `hasCartDomain` is true.)

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,219 @@
+		this.setState( { fields } );
+	};
+
+	addButton() {
```
Nit: for this function, `removeButton`, and `formFields`, I think it would be better to prefix them with `render` to make it clear that we're rendering the buttons, not performing the add/remove action for a button.
```suggestion
	renderAddButton() {
```

stephanethomas replied:
This `addButton()` is never called anywhere.

daledupreez replied:
I've renamed this function, but this is still a placeholder to support adding more than one email forward at a time. We're not trying to get that fully implemented in this PR, as the back end API only supports adding one forward at a time.

But I will update the PR description regardless.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,188 @@
+					/>
+				) : null }
+				{ this.state.forwards.map( ( fields, index ) => (
+					<Fragment key={ `fragment-${ index }` }>
```
Maybe call this something more unique? Same applies for the `form-` and `hr-` keys below.
```suggestion
					<Fragment key={ `email-forward-fragment-${ index }` }>
```

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,219 @@
+
+	formFields() {
+		const { translate, selectedDomainName, index, fields } = this.props;
+		const contactText = translate( 'contact', {
+			context: 'part of e-mail address',
+			comment: 'As it would be part of an e-mail address contact@example.com',
+		} );
+		const exampleEmailText = translate( 'e.g. %(example)s', {
+			args: { example: contactText },
+		} );
```
I think it may be better to have this be a single translate call so the context is not lost.

daledupreez replied:
I see we have this translated already using this pattern, so it may be better to keep it for now.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,219 @@
+import formState from 'calypso/lib/form-state';
+
+class EmailForwardingAddNewCompact extends Component {
+	static propTypes = {
```
As a high-level comment, this implementation is diverging from the approach we've used for the Google and Titan add new mailbox components. I'd love to know more about whether that was a conscious decision or more a function of replicating aspects of the existing email forwarding logic.

daledupreez replied:
I now see that this code is very similar to the existing `EmailForwardingAddNew` implementation, with all its approaches to the code!

AllTerrainDeveloper replied:
Yes, I tried to adapt the existing code but it ended yo modifying more than I expected. I'm open to keep using the existing code.

daledupreez commented on client/my-sites/email/email-forwarding/style.scss:
```diff
@@ -89,3 +89,13 @@ ul.email-forwarding__list {
 	margin-bottom: 15px;
 }
+
+.email-forwarding-add-new-compact-list__actions {
+	flex-direction: row;
+	display: flex;
+	justify-content: space-between;
+}
```
Should this be vertical for mobile devices?

AllTerrainDeveloper replied:
As right now there is no more buttons in this "row" it should be fine

daledupreez replied:
:+1:

daledupreez commented on client/my-sites/email/email-forwarding/style.scss:
```diff
@@ -89,3 +89,13 @@ ul.email-forwarding__list {
+	justify-content: space-between;
+}
+
+.gridicon.gridicons-trash {
+	margin-right: 4px;
+}
```
This selector needs to be more specific -- it needs to be wrapped by something otherwise it will apply to all of these icons.

AllTerrainDeveloper replied:
I think I can remove it for now too, is not possible to add/remove forwards now

daledupreez commented on client/my-sites/email/email-forwarding/index.jsx:
```diff
@@ -77,13 +82,22 @@ class EmailForwarding extends Component {
+					selectedDomainName={ selectedDomainName }
+				/>
+			);
+		}
```
Would it not be clearer to push this into its own component? Given that we eventually want to retire this code, I think it would be cleaner to just create a new component that pulls in the data we need.

daledupreez commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -299,12 +332,20 @@ class EmailProvidersComparison extends Component {
 			return null;
 		}
 
+		const formFields = (
+			<EmailForwarding
+				selectedDomainName={ selectedDomainName }
+				compact={ true }
+				onButtonClick={ this.goToEmailForwarding }
+			/>
+		);
+
```
Related to my other comments around re-using the existing component, is there a reason we're not just using the new component directly?

AllTerrainDeveloper replied:
My idea was to avoid regression with the existing component, that component is used later, after the first mailbox creations.

daledupreez replied:
We discussed this in a call. My issue here was that we should just be using `<EmailForwardingAddNewCompactList>` directly here, without making `EmailForwarding` support a new behaviour/purpose.

daledupreez commented on client/state/email-forwarding/actions.js:
```diff
@@ -42,12 +42,13 @@ export const receiveGetEmailForwardsFailure = ( domainName, error ) => {
 };
 
-export const addEmailForward = ( domainName, mailbox, destination ) => {
+export const addEmailForward = ( domainName, mailbox, destination, siteSlug = null ) => {
```
From an API perspective, I think it might be much easier to use a more explicit `onSuccessRedirectTarget` argument, so the caller can tell what the value will be used for rather than just specifying the site slug because it's easy to get.

AllTerrainDeveloper replied:
Much more elegant, I like it

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,188 @@
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination, siteSlug ) =>
+		withAnalytics(
+			composeAnalytics(
+				recordGoogleEvent(
+					'Domain Management',
+					'Clicked "Add New Email Forward" Button in Email Forwarding',
+					'Domain Name',
+					domainName
+				),
```
I don't believe we need this event -- Letero is not using any Google tracking.

AllTerrainDeveloper replied:
The original component for EmailForwarding does that also, would be nice to discuss a few things as you proposed :)

daledupreez replied:
We chatted about this -- we're not using Google Analytics elsewhere, so let's drop that tracking.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,219 @@
+				destination
+			);
+			this.formStateController.resetFields( this.getInitialFields() );
+			this.setState( { formSubmitting: false, showForm: true } );
```
This code assumes that `this.props.addNewEmailForwardWithAnalytics()` is synchronous. I don't think we can safely make that assumption. It would be better to have that tracking driven by data from Redux (assuming it is available).

daledupreez commented on client/my-sites/email/email-forwarding/index.jsx:
```diff
@@ -34,7 +36,10 @@ class EmailForwarding extends Component {
 	render() {
-		const { selectedDomainName, translate } = this.props;
+		const { compact = false, selectedDomainName, translate } = this.props;
+		if ( compact ) {
+			return this.renderContent();
+		}
```
This code path needs to make sure that we include `<QueryEmailForwards domainName={ selectedDomainName } />` to get `emailForwards` populated.

AllTerrainDeveloper replied:
But when we show this form, I think we are pretty sure that there aren't any other forwards for this domain, isn't it?

daledupreez replied:
Right now, we're only showing this form from within `EmailProvidersComparison`, but we plan to support using the form for adding new email forwards from a different context.
We should either have the component query for the email forwards, or ensure that the data is specified in the API for the component.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,188 @@
+		return (
+			<>
+				{ totalFowards > 0 ? (
+					<EmailForwardingLimit
+						emailForwardingCount={ totalFowards }
+						emailForwardingLimit={ emailForwardingLimit }
+					/>
```
The logic here is also counting the empty, unsubmitted form due to including `this.state.forwards.length` -- this looks a bit weird, because it adds 1 before I would expect it to.

AllTerrainDeveloper replied:
I think it's even better to remove it for now as we can only add one forward for the moment in this comparison page.

daledupreez reviewed (commented):
I have a number of comments on the PR -- it may be worth synching up in the morning to discuss how best to move forward on this.

AllTerrainDeveloper committed:
Emails: Amend pull requests suggestions.

AllTerrainDeveloper committed:
Emails: Lint issue

AllTerrainDeveloper committed:
Emails: Remove <EmailsForwards/> wrapper, rely on redux state for adding new forward, remove Google Analytics

AllTerrainDeveloper committed:
Emails: Lint issue

AllTerrainDeveloper committed:
Emails: Remove hasCartDomain condition

AllTerrainDeveloper committed:
Emails: Null pointer fix

daledupreez commented:
I also see that there is a legitimate unit test failure -- the tests need to be updated to handle the redirect handling.

AllTerrainDeveloper committed:
Emails: Fix tests

stephanethomas commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -1,14 +1,16 @@
-		return isValid ? [] : [ 'Invalid' ];
+		return isValid
+			? validateDuplicatedForward( { value, name: fieldName }, existingForwards )
+			: [ 'Invalid' ];
```
Not mandatory at all but here is another way to implement that logic that I find a bit easier to read:

```suggestion
		if ( ! isValid ) {
			return [ 'Invalid' ];
		}
		
		return validateDuplicatedForward( { value, name: fieldName }, existingForwards );
```

stephanethomas commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -23,6 +25,17 @@ function validateField( { name, value } ) {
+				: [];
+		default:
+			return [];
+	}
```
I think we should be able to simplify this block of code with:

```suggestion
	if ( name !== 'mailbox') {
		return [];
	}
	
	return existingForwards?.some( ( forward ) => forward.mailbox === value ) ? [ 'Repeated' ] : [];
```

stephanethomas commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -23,6 +25,17 @@ function validateField( { name, value } ) {
+	switch ( name ) {
+		case 'mailbox':
+			return existingForwards?.filter( ( t ) => t.mailbox === value ).length > 1
+				? [ 'Repeated' ]
```
What about `Duplicate` instead of `Repeated`? That would be more consistent with the name of that function.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+	}
+
+	validForwards() {
+		return ! this.state.forwards.some( ( t ) => ! t.valid );
```
We usually use a full word for variable names:

```suggestion
		return ! this.state.forwards.some( ( forward ) => ! forward.valid );
```

AllTerrainDeveloper replied:
Ups I'm using to this naming in functional programming

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
```
What about `maximumNumberOfEmailForwards` instead of `emailForwardingLimit`? This is slightly longer but it's also more expressive (otherwise it's difficult to understand what kind of limit is applied).

AllTerrainDeveloper replied:
I think we should stick with old name, just because there would be an incosistence between the getter (which is `getEmailForwardingLimit` ) and variable name. Don't you think so?

stephanethomas replied:
That's a good point indeed.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		return this.props.emailForwards.length >= this.props.emailForwardingLimit;
+	}
+
+	validForwards() {
```
I think this method should be named `hadValidForwards()` instead of `validForwards()`.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+			return;
+		}
+
+		this.state.forwards.map( ( t ) => {
```
Same comment about using full names for variables 🙂.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+	};
+
+	state = {
+		forwards: [ { destination: '', mailbox: '', valid: false } ],
```
I'm not a big fan of having some prop named `emailForwards` and some state named `forwards`. What's the difference between the two? What's the purpose of each of those data structure? Maybe we should rename `forwards` to `newEmailForwards` or something like that to reduce the confusion?

stephanethomas replied:
Oh, and `valid` should probably be renamed to `isValid` like with all boolean variables.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+
+	renderFormFooter() {
+		return <div>{ this.renderAddButton() }</div>;
+	}
```
It looks like we could remove this method, and just inline that code in the `render()` method.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		this.setState( { forwards: [ ...this.state.forwards, { destination: '', mailbox: '' } ] } );
+	};
+
+	renderAddButton( addMoreButton = false ) {
```
This is probably nitpicky but this method actually renders two buttons, so maybe we should find it a more appropriate name (e.g. `renderButtons` or `renderActionsButtons`).

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,185 @@
+		return this.props.fields;
+	}
+
+	UNSAFE_componentWillMount() {
```
We should no longer use `UNSAFE` methods like this one.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		return <div>{ this.renderAddButton() }</div>;
+	}
+
+	removeHandler = ( index ) => {
```
We could probably rename `removeHandler()` to `removeEmailForward()` as it is more descriptive. We also don't usually refer to event handlers with the `Handler` suffix. Ideally, we would also rename the corresponding prop (e.g. `onRemoveEmailForward` instead of `removeHandler`).

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		this.setState( { forwards: array } );
+	};
+
+	updateHandler = ( index, name, value ) => {
```
Same remark about renaming this method from `updateHandler()` to `updateEmailForward()`.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		const array = this.state.forwards;
+		array.splice( index, 1 );
+
+		this.setState( { forwards: array } );
```
I think destructuring could help here (and in the other handler):

```suggestion
		const [ forwards ] = this.state;
		forwards.splice( index, 1 );

		this.setState( { forwards } );
```

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		const array = this.state.forwards;
+		array[ index ][ name ] = value;
+
+		const valid = validateAllFields( array[ index ] );
```
I'm having a hard time understanding what this `valid` variable actually contains. Maybe it should be named `validEmailForward` or `emailForwardValidated`?

stephanethomas commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -23,6 +25,17 @@ function validateField( { name, value } ) {
+		default:
+			return [];
+	}
+}
```
Should we add this new function in a separate file (like for `getEmailForwardsCount()` and `hasEmailForwards()`)?

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,185 @@
+				return translate( 'Invalid mailbox - Duplicated' );
+			}
+		}
+		if ( fieldName === 'destination' ) {
```
Let's add blank lines before each of those `if` blocks 🙂.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+					<Fragment key={ `email-forwarding__add-new_fragment-${ index }` }>
+						<form className="email-forwarding__add-new" key={ `form-${ index }` }>
+							<EmailForwardingAddNewCompact
+								key={ `email-forwarding__add-ne-${ index }` }
```
Typo.

stephanethomas replied:
Actually I don't understand why we need a `key` here.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+		return (
+			<>
+				{ this.state.forwards.map( ( fields, index ) => (
+					<Fragment key={ `email-forwarding__add-new_fragment-${ index }` }>
```
Why do we need this fragment for? There is already a `key` on the `form` element anyway. Or maybe that should be the opposite, i.e. remove `key` from the `form` and `hr` elements.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+			addingForward: isAddingForward( state, ownProps.selectedDomainName ),
+			emailForwards: getEmailForwards( state, ownProps.selectedDomainName ),
+			emailForwardingLimit: getEmailForwardingLimit( state, getSelectedSiteId( state ) ),
+			emailForwardingType: getEmailForwardingType( state, ownProps.selectedDomainName ),
```
This prop is never used, and should be removed.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+
+		event.preventDefault();
+
+		if ( addingForward ) {
```
I understand why you didn't name `addingForward` as `isAddingForward` but that's something we could work around by not destructing this variable:

```suggestion
		if ( this.props.isAddingForward ) {
```

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,173 @@
+			const { mailbox, destination } = t;
+
+			this.addNewEmailForwardWithAnalytics(
+				this.props.selectedDomainName,
```
This variable could be destructured.

stephanethomas commented on client/state/selectors/get-email-forwards.js:
```diff
@@ -11,6 +11,17 @@ export function getEmailForwards( state, domainName ) {
+export function isAddingForward( state, domainName ) {
+	return state?.emailForwarding?.[ domainName ]?.addingForward ?? null;
+}
+
```
Why did we put this selector into its own file? I would also name it `isAddingEmailForward()` instead of `isAddingForward()`.

stephanethomas commented on client/state/selectors/get-email-forwards.js:
```diff
@@ -11,6 +11,17 @@ export function getEmailForwards( state, domainName ) {
 }
 
+/**
+ * Retrieves the state of the reducer for adding an email forward
```
I'm not sure we need to mention the reducer here:
```suggestion
 * Determines whether we are in the process of adding a new email forward to the specified domain.
```

stephanethomas commented on client/state/email-forwarding/reducer.js:
```diff
@@ -169,13 +182,32 @@ export const requestErrorReducer = ( state = false, action ) => {
 		requesting: requestingReducer,
 		requestError: requestErrorReducer,
+		addingForward: addingReducer,
+		addError: addErrorReducer,
```
Apologies if I've missed something but it doesn't look like this `addError` is used anywhere, so maybe we should remove it.

stephanethomas commented on client/state/email-forwarding/actions.js:
```diff
@@ -42,12 +44,18 @@ export const receiveGetEmailForwardsFailure = ( domainName, error ) => {
 		domainName,
 		mailbox,
 		destination,
+		onSuccessRedirectTarget,
```
I don't see why we refer to a 'target' here, I believe `onSuccessRedirect` or `onSuccessRedirectToPage` would probably more appropriate.

stephanethomas commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -314,11 +351,10 @@ class EmailProvidersComparison extends Component {
-				buttonLabel={ translate( 'Add email forwarding' ) }
 				showExpandButton={ this.isDomainEligibleForEmail( domain ) }
 				expandButtonLabel={ translate( 'Add email forwarding' ) }
-				onButtonClick={ this.goToEmailForwarding }
```
We should remove the corresponding `goToEmailForwarding()` method.

stephanethomas commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -240,17 +241,49 @@ class EmailProvidersComparison extends Component {
+	onForwardsFormReturnKeyPress = ( event ) => {
+		// Simulate form submission
+		if ( event.key === 'Enter' ) {
+			this.onForwardingConfirmNewMailboxes();
```
Here we have a `onForwards...()` method that simply calls a `onForwarding...()`. I'd love if we could stay consistent regarding naming, and use the same term everywhere.

stephanethomas replied:
Well, this `onForwardsFormReturnKeyPress()` is actually never used 😬.

stephanethomas commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -240,17 +241,49 @@ class EmailProvidersComparison extends Component {
 			} );
 	};
 
+	onForwardsMailboxesChange = ( changedMailboxes ) => {
```
I think this should be either `onEmailForwardsChange()` or `onForwardMailboxesChange()`.

stephanethomas replied:
Ok, this method is actually never used.

stephanethomas reviewed (changes requested):
I wasn't part of the initial discussion - so I'm certainly missing some context - but I find it a shame that we didn't use React Query nor TypeScript for those new components.

Anyway, I left plenty of comments that need to be addressed 🙂.

AllTerrainDeveloper committed:
Emails: PR Suggestions

AllTerrainDeveloper committed:
Emails: PR suggestions

AllTerrainDeveloper committed:
Emails: PR Suggestion

AllTerrainDeveloper committed:
Emails: Fix

AllTerrainDeveloper committed:
Emails: Linting

olaseni commented:
I haven't really looked at the code but there's something you need to address:

> Go to Upgrade -> Emails
Add a single email forward, click add. Check that the forward is done.
Assert that you have been redirected to forward management screen.

While there's a notice that the forward was created, the screen does not change to reveal the management screen.

<image src="https://user-images.githubusercontent.com/277661/140536626-69aa3a35-1099-45a5-942c-22759cb24648.jpg" width="900" />

AllTerrainDeveloper committed:
no message

olaseni commented on client/lib/domains/email-forwarding/has-duplicated-email-forwards.js:
```diff
@@ -0,0 +1,7 @@
+ * @param newEmailForward a string representing a new email forward
+ * @returns { boolean } If the email forward is present in the existing email forwards collection
+ */
+export function isDuplicatedEmailForwards( newEmailForward, existingEmailForwards ) {
```
I think it's a good idea to have the name of the file and function similar. That way you don't have to keep up with two contexts.

Said differently, might it be neater to call this function `hasDuplicatedEmailForwards()` ?

olaseni commented on client/state/email-forwarding/reducer.js:
```diff
@@ -169,13 +182,31 @@ export const requestErrorReducer = ( state = false, action ) => {
 	return state;
 };
 
+export const addErrorReducer = ( state = false, action ) => {
```
This isn't used, nor referred to. Let's remove it.

olaseni commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+		emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+	};
+
+	hasForwards() {
```
Since this is sort of describing `emailForwards`, it should probably be `hasEmailForwards`.

olaseni commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+		return this.props.emailForwards.length >= this.props.emailForwardingLimit;
+	}
+
+	hasValidForwards() {
```
This also can be `hasEmailForwards`

olaseni commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				{ addMoreButton && (
```
Seeing as we aren't supporting additional mailboxes right now, we should remove this.

olaseni commented on client/state/data-layer/wpcom/email-forwarding/add/index.js:
```diff
@@ -85,6 +87,8 @@ export const addEmailForwardSuccess = ( action, response ) => {
 			);
 		}
 
+		onSuccessRedirect();
```
Would it be simpler to add a reducer that monitors the success state and use that value to invoke a redirect in the component itself?

I'm wondering if having this here may be causing the issues with redirect not working.

olaseni commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -251,6 +241,27 @@ class EmailProvidersComparison extends Component {
 		}
 	};
 
+	onForwardingConfirmNewMailboxes = () => {
```
This is not called anywhere. Before you go ahead and remove this method, make plans to ensure we are sending the `calypso_email_providers_add_click` event with a valid `source` for forwards, so it can be consistent with what we do for Titan and Google.

olaseni reviewed (changes requested):
There are a bunch of unused methods and inert code paths that you really need to work on.

Also I'm able to replicate the `non-redirect on success` issue. I'll post vides. The issue might be related to where we are invoking the event callbacks.

olaseni commented:
> Also I'm able to replicate the `non-redirect on success` issue. 

The probability increases with increased network latency

https://user-images.githubusercontent.com/277661/140940567-0ae95a03-2a38-4f5a-84f6-b79402297ac7.mov

AllTerrainDeveloper committed:
Emails: Olaseni suggestions

AllTerrainDeveloper committed:
Emails: Remove some warnings from console, add Olaseni suggestion for redirect.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,159 @@
+	};
+
+	onUpdateEmailForward = ( index, name, value ) => {
+		const array = this.state.emailForwards;
+		array[ index ][ name ] = value;
```
Two nits:
1. I would find this code easier to understand if we called the variable `emailForwards` instead of `array`.
2. I don't like that we're using a `const` and then updating a value on the next line. The linter may not allow us to use a `let`, but I'd prefer that, as we are modifying the variable.

AllTerrainDeveloper replied:
1. I'll rename :)
2. about const, I used const because we are not really modifying the variable (not reassigning it's value eg. memory address/pointer), but modifying it's content. Anyway I get your point, and it's also weird to do some modification on the next line.

daledupreez commented on client/my-sites/email/email-forwarding/style.scss:
```diff
@@ -89,3 +89,9 @@ ul.email-forwarding__list {
 	line-height: 3em;
 	margin-bottom: 15px;
 }
+
+.email-forwarding-add-new-compact-list__actions {
+	flex-direction: row;
+	display: flex;
+	justify-content: space-between;
+}
```
I think this is the wrong place for this style block -- this `style.scss` file is only imported by `EmailForwarding`, and is not imported by `EmailForwardingAddNewCompactList.jsx`, so the styles are not going to be applied when the latter component is imported from the email comparison page.

daledupreez commented on client/state/data-layer/wpcom/email-forwarding/add/index.js:
```diff
@@ -62,7 +64,7 @@ export const addEmailForwardFailure = ( action, error ) => {
 
 export const addEmailForwardSuccess = ( action, response ) => {
-	const { domainName, mailbox, destination } = action;
+	const { domainName, mailbox, destination, onSuccessRedirect = noop } = action;
```
It doesn't look like we're supplying this property any more. If so, I think we should remove it and the related handling.

daledupreez commented on client/state/email-forwarding/actions.js:
```diff
@@ -42,12 +44,13 @@ export const receiveGetEmailForwardsFailure = ( domainName, error ) => {
 };
 
-export const addEmailForward = ( domainName, mailbox, destination ) => {
+export const addEmailForward = ( domainName, mailbox, destination, onSuccessRedirect = noop ) => {
 	return {
 		type: EMAIL_FORWARDING_ADD_REQUEST,
 		domainName,
 		mailbox,
 		destination,
+		onSuccessRedirect,
```
It also looks like the `onSuccessRedirect` argument is not needed here.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,159 @@
+	render() {
+		const { emailForwardSuccess, selectedDomainName, selectedSiteSlug } = this.props;
+
+		if ( emailForwardSuccess ) {
+			page( emailManagement( selectedSiteSlug, selectedDomainName ) );
+		}
```
Please `return` after this line -- this may be triggering the weird issues 'Seni reported.

Actually, I am also worried that this is the wrong place for the redirect logic. This component could be added anywhere in the UI, so redirecting to the email management page doesn't belong in this component.

I think it would be better to set up an optional `onSuccessRedirect` prop to handle this better.

```suggestion
		// We could also use something like `onSuccessRedirectHandler` as a function, but I don't know if it's worth generalising that much
		if ( emailForwardSuccess && onSuccessRedirectDestination ) {
			page( onSuccessRedirectDestination );
			return;
		}
```

AllTerrainDeveloper replied:
I've also took advice from react itself, it said in the console, better to move that code to componentDidUpdate

daledupreez reviewed (commented):
I haven't tested yet, but I have some concerns that should be addressed.

AllTerrainDeveloper committed:
Emails: PR suggestions
Renamed array to emailsForwards
Cleared console errors (discount expected to be object)
Cleared console errors (React advised to move redirect code to componentDidUpdate)
Remove unnecessary code

AllTerrainDeveloper committed:
Remove unnecessary style

AllTerrainDeveloper committed:
Emails: Use let instead of const

AllTerrainDeveloper commented:
I've also fixed a console error unrelated to this PR, discount should be object in prototypes, instead of string

daledupreez commented on client/lib/domains/email-forwarding/has-duplicated-email-forwards.js:
```diff
@@ -0,0 +1,7 @@
+ * @returns { boolean } If the email forward is present in the existing email forwards collection
+ */
+export function hasDuplicatedEmailForwards( newEmailForward, existingEmailForwards ) {
+	return existingEmailForwards?.some( ( forward ) => forward.mailbox === newEmailForward );
```
Nit: It may be worth supplying a boolean for the fall-through case where `existingEmailForwards` is not an array. (Can you tell I've been writing TypeScript recently? ;) )
```suggestion
	return existingEmailForwards?.some( ( forward ) => forward.mailbox === newEmailForward ) ?? false;
```

daledupreez commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -1,14 +1,23 @@
 
-function validateAllFields( fieldValues ) {
+function validateAllFields( fieldValues, existingForwards = [] ) {
 	return mapValues( fieldValues, ( value, fieldName ) => {
```
Not to future us: it would be good to refactor this code away from `lodash` when we get a chance.

That's not something we should do in this PR, though!

daledupreez commented on client/my-sites/email/email-providers-comparison/email-provider-card.jsx:
```diff
@@ -114,7 +114,7 @@ EmailProviderCard.propTypes = {
 	badge: PropTypes.object,
 	description: PropTypes.string,
 	formattedPrice: PropTypes.node,
-	discount: PropTypes.string,
```
Not blocking, but it might have been easier to handle this in a separate PR.

AllTerrainDeveloper replied:
I'll revert :) I'll create Asana card too, as there is an ugly error in console.

daledupreez commented on client/state/selectors/get-email-forwards.js:
```diff
@@ -11,6 +11,28 @@ export function getEmailForwards( state, domainName ) {
+ * @returns {boolean}          adding forwards
+ */
+export function isAddingEmailForward( state, domainName ) {
+	return state?.emailForwarding?.[ domainName ]?.addingForward ?? null;
```
Should we not default to `?? false` given that we're declaring the return value to be a `boolean`?
```suggestion
	return state?.emailForwarding?.[ domainName ]?.addingForward ?? false;
```

daledupreez commented on client/my-sites/email/email-providers-comparison/style.scss:
```diff
@@ -228,6 +228,20 @@
 			flex-grow: 1;
 			margin-bottom: 20px;
 
+			.is-compact {
+				padding: 0;
+			}
+
+			.email-forwarding__list {
+				margin-top: 0;
+			}
```
I don't believe either of these selectors are in use -- I think both are stale.

daledupreez reviewed (commented):
I have mostly minor comments, but when I tested, I didn't get redirected. I will debug that more in the morning to work out what is going on.

AllTerrainDeveloper commented:
Hmmm I'll debug. It redirected for me 🙃

AllTerrainDeveloper committed:
Emails: Forward fixes.
Fetch domains after forward has been added form email/management screen.
Revert PropType.object to PropType.string (will be addressed in different PR)
Remove unnecessary styling
Remove unnecessary code (are more forwards)
Add guards to ONLY retrieve domains once, do not redirect twice or automatically
Fallback to false instead of null in "isAddingEmailForward"

AllTerrainDeveloper commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,176 @@
+			siteId,
+		} = this.props;
+
+		if (
+			emailForwardSuccess &&
+			onSuccessRedirectDestination &&
+			! this.state.isRedirecting &&
+			this.state.newForwardAdded
+		) {
+			if ( ! isRequestingDomains && ! this.state.isRedirecting ) {
+				this.props.fetchSiteDomains( siteId );
+			}
+			page( onSuccessRedirectDestination );
+			this.state.isRedirecting = true;
+		}
```
All this guards are to prevent the following cases:

- We have added a forward on a previous step, so emailForwardSuccess is already true, so in case we are in purchase page, we are redirected automatically to email/manage!
- We don't redirect twice while redirecting
- We don't fetch while fetching domains
- We don't redirect without having previously added a new forward (newForwardAdded)

AllTerrainDeveloper replied:
All this guards are to prevent the following cases:

- We have added a forward on a previous step, so emailForwardSuccess is already true, so in case we are in purchase page, we are redirected automatically to email/manage!
- We don't redirect twice (or more) while redirecting
- We don't fetch while fetching domains
- We don't redirect without having previously added a new forward (newForwardAdded)

PS: I couldn't add fetchSiteDomains in the deconstruction as it shadows import

AllTerrainDeveloper committed:
Emails: Deconstruct state

AllTerrainDeveloper commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,178 @@
+		this.setState( { emailForwards } );
+	};
+
+	componentDidUpdate() {
+		const {
+			emailForwardSuccess,
+			isRequestingDomains,
+			onSuccessRedirectDestination,
+			siteId,
+		} = this.props;
+
+		const { isRedirecting, newForwardAdded } = this.state;
+
+		if (
+			emailForwardSuccess &&
+			onSuccessRedirectDestination &&
+			! isRedirecting &&
+			newForwardAdded
+		) {
+			if ( ! isRequestingDomains && ! isRedirecting ) {
+				this.props.fetchSiteDomains( siteId );
+			}
+			page( onSuccessRedirectDestination );
+			this.state.isRedirecting = true;
+		}
+	}
```
All this guards are to prevent the following cases:

- We have added a forward on a previous step, so emailForwardSuccess is already true, so in case we are in purchase page, we are redirected automatically to email/manage!
- We don't redirect twice (or more) while redirecting
- We don't fetch while fetching domains
- We don't redirect without having previously added a new forward (newForwardAdded)

PS: I couldn't add fetchSiteDomains in the deconstruction as it shadows import

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,178 @@
+			! isRedirecting &&
+			newForwardAdded
+		) {
+			if ( ! isRequestingDomains && ! isRedirecting ) {
```
I am on the fence about whether this code should always trigger the domains fetch. I have two concerns:
* the receiving component should generally be ensuring that it has the correct data
* do we always need to trigger a fetch? Maybe it’s only needed for the first one. (I am not sure about this, to be honest.)

AllTerrainDeveloper replied:
I am not sure what to do here. I had in mind that just after adding new forwards, our store would already have updated data. As this wasn't the case, we need to update our data as soon as we have added a new forward, so this is the case. Also I see this fetch a bit "anti pattern" but I can't think a better way to do so.

We could fetch always in the emails/home or emails/management but I think it's not nice to fetch if we haven't change anything in the client.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,178 @@
+				this.props.fetchSiteDomains( siteId );
+			}
+			page( onSuccessRedirectDestination );
+			this.state.isRedirecting = true;
```
Please call `setState()` rather than setting the value directly.

AllTerrainDeveloper replied:
As discussed in Slack, is not possible, as we would trigger componentDidUpdate forever 😢. I think we can remove this guard (there should not be any problem re-redirecting), or keep using it, as its the end of life-cycle for this component.

daledupreez reviewed (commented):
I have some comments from inspection, but haven’t been able to test on my phone - calypso.live is not happy.

AllTerrainDeveloper committed:
Emails: Final code? - Refetch domains

AllTerrainDeveloper committed:
Emails: Unused import

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onSuccessAddForwarding: PropTypes.func,
```
Nit: not sure why, but I would prefer this to be `onAddForwardSuccess` or `onAddForwardingSuccess`, as that uses pattern more like onActionResult. If we had a similar failure, we'd then have that called something like `onAddForwardFailure` and the two values would be next to each other.

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+				selectedSiteSlug
+			);
+		} );
+		this.state.newForwardAdded = true;
```
This should be a `setState()` call.

daledupreez committed:
Rename prop/callback to onAddForwardSuccess

daledupreez committed:
Use setState()

daledupreez commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+	};
+
+	componentDidUpdate( prevProps ) {
+		const { emailForwardSuccess, onSuccessAddForwarding } = this.props;
+
+		const { newForwardAdded } = this.state;
+
+		if (
+			emailForwardSuccess &&
+			onSuccessAddForwarding &&
+			newForwardAdded &&
+			prevProps.emailForwardSuccess !== emailForwardSuccess
+		) {
+			onSuccessAddForwarding();
+		}
```
Neatly done here! This was really subtle and using `prevProps` makes a huge difference to understanding when the success flag changes.

daledupreez commented on client/my-sites/email/email-providers-comparison/index.jsx:
```diff
@@ -298,13 +299,29 @@ class EmailProvidersComparison extends Component {
 
+	onSuccessAddForwarding = () => {
+		const { domain, getSiteDomains, requestingSiteDomains, selectedSite } = this.props;
+		if ( ! requestingSiteDomains ) {
+			getSiteDomains( selectedSite.ID );
+		}
```
For now, I think this is OK.

I just took a closer look and the issue seems to be that we're not adding a `<QuerySiteDomains />` component to the content we render in `EmailManagementHome` for the two code paths where we end up rendering `EmailProvidersComparison`. But we can take that on in a follow-up PR.

daledupreez reviewed (commented):
Well done on sticking with this to get it across the finish line!

I am going to tackle my own comments on two small items, and will approve the PR once I've retested.

daledupreez committed:
Don't show email forward warning before we redirect to management page

daledupreez reviewed (commented):
While I was testing , I noticed that the code is still triggering the notice that @olaseni reported in [this comment](https://github.com/Automattic/wp-calypso/pull/57492#issuecomment-961995133). We have something of a timing gap between the Add button being pressed and the new page loading, during which time we see the notice about existing email forwards. I've tried to handle that in commit 61d1326.

But I am going to need someone to kick the tires tomorrow.

stephanethomas commented on client/lib/domains/email-forwarding/index.js:
```diff
@@ -1,14 +1,23 @@
+import { hasDuplicatedEmailForwards } from 'calypso/lib/domains/email-forwarding/has-duplicated-email-forwards';
 
-function validateAllFields( fieldValues ) {
+function validateAllFields( fieldValues, existingForwards = [] ) {
```
Minor but it would be great to be consistent regarding naming (in this particular case, use the same term than in the `hasDuplicatedEmailForwards()` function:

```suggestion
function validateAllFields( fieldValues, existingEmailForwards = [] ) {
```

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onAddForwardSuccess: PropTypes.func,
```
Same remark about naming:

```suggestion
		onAddEmailForwardSuccess: PropTypes.func,
```

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
+			newForwardAdded: false,
```
Same thing:

```suggestion
			newEmailForwardAdded: false,
```

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+
+		event.preventDefault();
+
+		if ( this.props.isAddingEmailForward ) {
```
We could destructure that variable.

daledupreez replied:
I ended up renaming the prop and destructuring because the previous name clashed with an imported function.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,167 @@
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
+		} );
+	};
```
This method is called anywhere, and should also be renamed.

stephanethomas commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact.jsx:
```diff
@@ -0,0 +1,188 @@
+			<div className="email-forwarding-add-new-compact__actions">
+				<Button
+					className="email-forwarding-add-new-compact__add-another-forward-button"
+					onClick={ this.onForwardAdd }
```
There is no event handler named `onForwardAdd()` in this component.

daledupreez replied:
I've updated this code to call a function supplied via a prop.

stephanethomas reviewed (changes requested):
I've found a few blocking issues. I didn't perform a full review as it seems some code was refactored but either not tested, or incorrectly rebased.

> Add a duplicated mailbox

I'm not sure how to actually test that, since it's not yet possible to add two email forwards at once. Moreover, once an email forward has been created, we are no longer presented with the `Email Comparison` page.

Finally, there are a few minor things that need to be addressed but this can happen in another pull request:

![image](https://user-images.githubusercontent.com/594356/141986613-e424c233-ce62-465d-8ea5-354b69edb518.png)

1. There is a light gray line just above the `Emails Sent To` label
2. The `Add` button should be primary (it should be colored)
3. All input fields labels should be lowercase, except for the first word
4. The margin between the last input field and the line should be increased to be consistent with other forms

daledupreez committed:
Use existingEmailForwards in validateAllFields()

daledupreez committed:
Rename onAddForwardSuccess to onAddEmailForwardSuccess

daledupreez committed:
Rename newForwardAdded to newEmailForwardAdded

daledupreez committed:
Rename isAddingEmailForward prop to isSubmittingEmailForward

daledupreez committed:
More renaming for add new forward button and handlers

daledupreez committed:
Small style fixes

AllTerrainDeveloper commented:
I've updated screenshot to show the subtle changes in styling (border top + margin between button and 

> I've found a few blocking issues. I didn't perform a full review as it seems some code was refactored but either not tested, or incorrectly rebased.
> 
> > Add a duplicated mailbox
> 
> I'm not sure how to actually test that, since it's not yet possible to add two email forwards at once. Moreover, once an email forward has been created, we are no longer presented with the `Email Comparison` page.
> 
> Finally, there are a few minor things that need to be addressed but this can happen in another pull request:
> 
> ![image](https://user-images.githubusercontent.com/594356/141986613-e424c233-ce62-465d-8ea5-354b69edb518.png)
> 
> 1. There is a light gray line just above the `Emails Sent To` label
> 2. The `Add` button should be primary (it should be colored)
> 3. All input fields labels should be lowercase, except for the first word
> 4. The margin between the last input field and the line should be increased to be consistent with other forms

About the colored button, it is colored only when the form is valid, right now is "disabled" state so it's greyish. I can remove that "feature" but then not sure if it will be good for the user to click on the button and see for him that nothing happens.

daledupreez commented:
> About the colored button, it is colored only when the form is valid, right now is "disabled" state so it's greyish. I can remove that "feature" but then not sure if it will be good for the user to click on the button and see for him that nothing happens.

I looked at handling this last night, but the click handling wasn't checking for validity, so I skipped that for this PR. The core issue is that the UX for the other forms on this page always show the button as enabled, and show any validation errors that prevent submission.

> 3. All input fields labels should be lowercase, except for the first word

This is due to the original form having these capitalised strings. We kept the original case to avoid re-translation (at least for now).

AllTerrainDeveloper commented:
It looks good to me all those changes!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6984410

Thank you @AllTerrainDeveloper for including a screenshot in the description! This is really helpful for our translators.

tyxla commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
```
The `isRedirecting` component state property isn't used anywhere and can be removed.

tyxla commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	onRemoveEmailForward = ( index ) => {
+		const [ emailForwards ] = this.state;
```
This looks like a mistake, `this.state` is an object; did you mean:

```js
const { emailForwards } = this.state;
```

tyxla commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+
+	onAddNewEmailForward = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
```
Updating the state of a component based on the current value of `this.state` may potentially lead to inconsistent component state.

This is also documented in the [React docs](https://reactjs.org/docs/react-component.html#setstate):

> `setState()` does not always immediately update the component. It may batch or defer the update until later. This makes reading `this.state` right after calling `setState()` a potential pitfall. Instead, use `componentDidUpdate` or a `setState` callback (`setState(updater, callback)`)

The easiest way to fix this is to just change `setState()` to the callback form, where it will always properly access the current state value.

tyxla commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+	onUpdateEmailForward = ( index, name, value ) => {
+		// eslint-disable-next-line prefer-const
+		let emailForwards = this.state.emailForwards;
+		emailForwards[ index ][ name ] = value;
```
We're directly mutating `this.state` here, and that's forbidden. Again, as per the [docs](https://reactjs.org/docs/react-component.html#setstate):

> `state` is a reference to the component state at the time the change is being applied. It should not be directly mutated.

To fix this, you should use a copy of `this.state.emailForwards` instead of referencing the original object, for example:

```
let emailForwards = { ...this.state.emailForwards };
```

tyxla commented on client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx:
```diff
@@ -0,0 +1,171 @@
+		emailForwards[ index ][ name ] = value;
+
+		const validEmailForward = validateAllFields( emailForwards[ index ] );
+		emailForwards[ index ].isValid =
```
This is also `this.state` mutation. Will also get fixed with the suggestion above.

tyxla reviewed (commented):
Sorry for the late review, but I've noticed several bugs in this PR. 

Would you mind following up with a PR to address them?

Let me know if you need anything or if you have any questions!

Thanks a bunch! 🙌

daledupreez commented:
Thanks for the review, @tyxla!

I'm tracking the follow-up work here: 1200182182542585-as-1201407613222688/f

tyxla commented:
Thanks, @daledupreez! Feel free to ping me if you need anything!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6984410).
